src/man1/qpwgraph.1
.TH qpwgraph "1" "April 16, 2026"
.SH NAME
qpwgraph \- A PipeWire Graph Qt GUI Interface
.SH SYNOPSIS
.B qpwgraph
[\fIoptions\fR] [\fIpatchbay-file\fR]
.SH DESCRIPTION
This manual page documents briefly the
.B qpwgraph
command-line options.
.PP
\fBqpwgraph\fP is a graph manager dedicated to PipeWire (https://pipewire.org),
using the Qt C++ framework (https://qt.io), based and pretty much like the same of
QjackCtl (https://qjackctl.sourceforge.io).
.PP
Source code repository: https://gitlab.freedesktop.org/rncbc/qpwgraph
.SH OPTIONS
.HP
\fB\-a\fR, \fB\-\-activated\fR
.IP
Activated patchbay.
.HP
\fB\-d\fR, \fB\-\-deactivated\fR
.IP
Deactivated patchbay.
.HP
\fB\-x\fR, \fB\-\-exclusive\fR
.IP
Exclusive patchbay.
.HP
\fB\-n\fR, \fB\-\-nonexclusive\fR
.IP
Non-exclusive patchbay.
.HP
\fB\-m\fR, \fB\-\-minimized\fR
.IP
Start minimized.
.HP
\fB\-r\fR, \fB\-\-remote\fR <\fIname\fR>
.IP
Remote daemon name.
.HP
\fB\-?\fR, \fB\-\-help\fR
.IP
Displays help on command-line options.
.HP
\fB\-v\fR, \fB\-\-version\fR
.IP
Show version information.
.SH FILES
Configuration settings are stored in ~/.config/rncbc.org/qpwgraph.conf
.SH AUTHOR
qpwgraph was written by Rui Nuno Capela.


ChangeLog
qpwgraph - A PipeWire Graph Qt GUI Interface
--------------------------------------------

ChangeLog


1.0.1  2026-04-21  A Spring'26 Release.

- Added remote daemon name and modified state labels to the main
  status-bar.

- Added command line option to connect on a remote PipeWire
  daemon context, by name (-r, --remote); nb. PIPEWIRE_REMOTE
  environment variable takes precedence though.

- Separate the singleton/unique application instance setup logic
  from system-tray icon build option; add PipeWire remote-name,
  if any, to singleton/unique application instance identifier.

- Settling on all boards to next development cycle (Qt >= 6.11)


1.0.0  2026-03-17  An Early-Spring'26 Release.

- Configured ALSA MIDI port colors now reset/restored properly.


0.9.9  2026-02-10  A Mid-Winter'26 Release.

- Most probably the first and last release candidate ever ;)


0.9.8  2025-12-29  An End-of-Year'25 Beta Release.

- Fixed 'mime-info' file to the correct MIME-type icon names.


0.9.7  2025-11-14  An Autumn'25 Beta Release.

- Add topological sort node arrangement (cf. View/Arrange Nodes;
  by Mike Bourgeous)


0.9.6  2025-10-06  An Early-Fall'25 Beta Release.

- Allow the complete node name to get the same treatment in the
  Graph/Options.../Filter as same for Merger.

- Get rid of CONFIG_WAYLAND build config option; add underlying
  platform name (eg. xcb, wayland) to Qt version string.


0.9.5  2025-08-15  A Mid-Summer'25 Beta Release.

- Make all node-names in merger list (Graph/Options.../Merger)
  implicitly exclusive for Patchbay persistence.

- Cope with in-flight node/port count changing.


0.9.4  2025-06-21  An Early-Summer'25 Beta Release.

- Indulged on a new 'Add' (pinned connection) button into the
  Patchbay/Manage... dialog also make dialog size and position
  persistent.
- Introducing Graph/Options.../Merger to unify node-names for
  Patchbay persistence, especially useful to PipeWire clients
  that spawn more than one node, having the very same name (eg.
  web browsers).


0.9.3  2025-06-02  An End-of-Spring'25 Beta Release.

- Corrected (again) salvage of node positions that possible have
  the very same and exact name.


0.9.2  2025-05-17  A Mid-Spring'25 Hot-Fix Beta Release.

- Save node positions as soon as they are taken or moved around,
  avoiding accidental re-positioning to the previous or original
  state.

- Regression to the late uniform and generic patchbay node-names
  which was a terrible mistake: while it sort of fixed some issues
  regarding multimedia players, it sent all other cases bonkers.
  So sorry for all the foul play or inconvenience.


0.9.1  2025-05-15  A Mid-Spring'25 Beta Release.

- Patchbay node-name persistence is now made uniform and generic
  for most if not all multimedia players (eg. web browsers) that
  usually tend to power-cycle their PipeWire client/driver on a
  whim (eg. pause, skip,... whatever)


0.9.0  2025-04-25  A Spring'25 Beta Release.

- Mitigate an alternate re-positioning to some application nodes
  (eg. Discord) on each screen refresh (F5).
- Stepping up next development cycle (Qt >= 6.9)


0.8.3  2025-03-28  An Early Spring'25 Beta Release.

- Loading a patchbay profile (*.qpwgraph) from the command line,
  is now properly shown on the main window title and toolbar.
- Added PipeWire command line version information.


0.8.2  2025-03-07  An End-of-Winter'25 Beta Release.

- Port colors: new PipeWire MIDI 2 (UMP) port type is now aliased
  to the same as (now old) MIDI port type. (pipewire >= 1.3.81)
- Fixed command line parsing (QCommandLineParser/Option) to not
  exiting the application with a segfault when showing help and
  version information.


0.8.1  2024-12-27  An End-of-Year'24 Beta Release.

- Patchbay 'presets' may now be selected from the system-tray icon
  context-menu.


0.8.0  2024-11-14  A Mid-Autumn'24 Beta Release.

- A new Patchbay/Manage... feature dialog is introduced, targeting
  the currently loaded patchbay, allowing the removal and cleanup
  of connection rules that are no longer used, obsolete or simply
  not applicable anymore.
- Force an actual complete graph refresh on main View/Refresh...
- Introducing the new Graph/Options... dialog, to where the most
  general option settings have been (re)moved from the main Help
  menu.
- Introducing Graph/Options.../Filter to hide nodes by node name
  pattern (blacklist).
- Fix default PipeWire node nickname if given empty.


0.7.9  2024-10-28  An Autumn'24 Beta Release.

- Thumb-view: Fixed flicker when dragging widget from top to
  bottom corners (and vice-versa); inner thumb-viewport is
  drag+moved only if the Ctrl keyboard modifier is pressed.
- Node reference positioning changed to the top-left corner,
  improving the base snapping-to-grid perception.
- Prepping up next development cycle (Qt >= 6.8)


0.7.8  2024-09-19  An End-of-Summer'24 Beta Release.

- When visible the thumb-view may now be drag-moved to a different
  corner position anytime.
  

0.7.7  2024-08-21  A Mid-Summer'24 Beta Release.

- Thumb-view repositions and resizes immediately when visible.


0.7.5  2024-07-12  A Summer'24 Beta Release.

- Implement patchbay node search functionality: adds a floating
  text-entry field, which appears when typing in the canvas,
  selecting nodes that contains the typed text (maybe regular
  expression).


0.7.4  2024-06-28  An Early-Summer'24 Hot-fix Release.

- Force a default PipeWire node name if given empty.
- Fixed a potential use-after-free issue on canvas nodes refresh
  and recycle.


0.7.3  2024-06-22  An Early-Summer'24 Beta Release.

- Cancel button option added to close to system-tray icon message.
- Introducing thumbview context-menu.


0.7.2  2024-05-12  A Mid-Spring'24 Beta Release.

- Remove margins and spacing from the top-level canvas layout.
- Possible fix to pipewire backend thread loop locking.
- Updated to latest framework level (Qt >= 6.7)


0.7.1  2024-04-25  A Spring'24 Beta Release Hot-Fix.


src/qpwgraph.cpp
// qpwgraph.cpp
//
/****************************************************************************
   Copyright (C) 2021-2026, rncbc aka Rui Nuno Capela. All rights reserved.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

*****************************************************************************/

#include "qpwgraph.h"
#include "qpwgraph_main.h"

#include <pipewire/pipewire.h>

#include <QTextStream>

#include <QCommandLineParser>
#include <QCommandLineOption>

#ifdef CONFIG_XUNIQUE
#include <QSharedMemory>
#if QT_VERSION >= QT_VERSION_CHECK(6, 6, 0)
#include <QNativeIpcKey>
#endif
#include <QLocalServer>
#include <QLocalSocket>
#include <QHostInfo>
#endif


//-------------------------------------------------------------------------
// Singleton application instance - impl.
//

// Constructor.
qpwgraph_application::qpwgraph_application ( int& argc, char **argv )
	: QApplication(argc, argv), m_widget(nullptr)
#ifdef CONFIG_XUNIQUE
	, m_memory(nullptr), m_server(nullptr)
#endif
	, m_patchbay_activated(-1)
	, m_patchbay_exclusive(-1)
	, m_start_minimized(false)
	, m_remote_name("pipewire-0")
{
	QApplication::setApplicationName(PROJECT_NAME);
	QApplication::setApplicationDisplayName(PROJECT_DESCRIPTION);

	QApplication::setDesktopFileName(
		QString("org.rncbc.%1").arg(PROJECT_NAME));

	QApplication::setApplicationVersion(PROJECT_VERSION);
}


// Destructor.
qpwgraph_application::~qpwgraph_application (void)
{
#ifdef CONFIG_XUNIQUE
	clearServer();
#endif
}


// Parse command line arguments.
bool qpwgraph_application::parse_args ( const QStringList& args )
{
	QCommandLineParser parser;
	parser.setApplicationDescription(
		PROJECT_NAME " - " + QObject::tr(PROJECT_DESCRIPTION));

	const QString s_activated    = "activated";
	const QString s_deactivated  = "de" + s_activated;
	const QString s_exclusive    = "exclusive";
	const QString s_nonexclusive = "non" + s_exclusive;
	const QString s_minimized    = "minimized";
	const QString s_remote_name  = "remote";
	const QString s_help         = "help";

	parser.addOption({{"a", s_activated},
		QObject::tr("Activated patchbay.")});
	parser.addOption({{"d", s_deactivated},
		QObject::tr("Deactivated patchbay.")});
	parser.addOption({{"x", s_exclusive},
		QObject::tr("Exclusive patchbay.")});
	parser.addOption({{"n", s_nonexclusive},
		QObject::tr("Non-exclusive patchbay.")});
	parser.addOption({{"m", s_minimized},
		QObject::tr("Start minimized.")});
	parser.addOption({{"r", s_remote_name},
		QObject::tr("Remote daemon name."),
		QObject::tr("name")});
	parser.addOption({{"?", s_help},
		QObject::tr("Displays help on command-line options.")});
	const QCommandLineOption& versionOption = parser.addVersionOption();
	parser.addPositionalArgument("patchbay-file",
		QObject::tr("Patchbay file (.%1)")
			.arg(QString(PROJECT_NAME).toLower()),
		QObject::tr("[patchbay-file]"));

	QTextStream out(stderr);

	if (!parser.parse(args)) {
		out << parser.errorText() << '\n';
		return false;
	}

	if (parser.isSet(s_help)) {
		out << parser.helpText() << '\n';
		return false;
	}

	if (parser.isSet(versionOption)) {
		out << QString("%1 %2\n")
			.arg(PROJECT_NAME)
			.arg(QCoreApplication::applicationVersion());
		out << QString("Qt: %1").arg(qVersion());
	#if defined(QT_STATIC)
		out << "-static";
	#endif
	#if QT_VERSION >= QT_VERSION_CHECK(5, 5, 0)
		out << ' ' << '(';
		out << QApplication::platformName();
		out << ')';
	#endif
		out << '\n';
		out << QString("PipeWire: %1 (headers: %2)")
			.arg(pw_get_library_version())
			.arg(pw_get_headers_version());
		out << '\n' << '\n';
		return false;
	}

	if (parser.isSet(s_activated))
		m_patchbay_activated = 1;
	else
	if (parser.isSet(s_deactivated))
		m_patchbay_activated = 0;

	if (parser.isSet(s_exclusive))
		m_patchbay_exclusive = 1;
	else
	if (parser.isSet(s_nonexclusive))
		m_patchbay_exclusive = 0;

	m_start_minimized = parser.isSet(s_minimized);

	if (parser.isSet(s_remote_name))
		m_remote_name = parser.value(s_remote_name);

	int nargs = 0;
	m_patchbay_path.clear();
	foreach (const QString& arg, parser.positionalArguments()) {
		if (nargs > 0)
			m_patchbay_path += ' ';
		m_patchbay_path += arg;
		++nargs;
	}

	// Alright with argument parsing.
	return true;
}


#ifdef CONFIG_XUNIQUE

// Check if another instance is running,
// and raise its proper main widget...
bool qpwgraph_application::setupServer (void)
{
	clearServer();

	m_unique = QCoreApplication::applicationName();
	QString uname = QString::fromUtf8(::getenv("USER"));
	if (uname.isEmpty())
		uname = QString::fromUtf8(::getenv("USERNAME"));
	if (!uname.isEmpty()) {
		m_unique += ':';
		m_unique += uname;
	}
	QString uremote = QString::fromUtf8(::getenv("PIPEWIRE_REMOTE"));
	if (uremote.isEmpty())
		uremote = m_remote_name;
	if (!uremote.isEmpty()) {
		m_unique += ':';
		m_unique += uremote;
		m_remote_name = uremote;
	}
	m_unique += '@';
	m_unique += QHostInfo::localHostName();
#if QT_VERSION >= QT_VERSION_CHECK(6, 6, 0)
	const QNativeIpcKey native_key
		= QSharedMemory::legacyNativeKey(m_unique);
	m_memory = new QSharedMemory(native_key);
#else
#if defined(Q_OS_UNIX)
	m_memory = new QSharedMemory(m_unique);
	m_memory->attach();
	delete m_memory;
#endif
	m_memory = new QSharedMemory(m_unique);
#endif
	bool is_server = false;
	const qint64 pid = QCoreApplication::applicationPid();
	struct Data { qint64 pid; };
	if (m_memory->create(sizeof(Data))) {
		m_memory->lock();
		Data *data = static_cast<Data *> (m_memory->data());
		if (data) {
			data->pid = pid;
			is_server = true;
		}
		m_memory->unlock();
	}
	else
	if (m_memory->attach()) {
		m_memory->lock(); // maybe not necessary?
		Data *data = static_cast<Data *> (m_memory->data());
		if (data)
			is_server = (data->pid == pid);
		m_memory->unlock();
	}

	if (is_server) {
		QLocalServer::removeServer(m_unique);
		m_server = new QLocalServer();
		m_server->setSocketOptions(QLocalServer::UserAccessOption);
		m_server->listen(m_unique);
		QObject::connect(m_server,
			SIGNAL(newConnection()),
			SLOT(newConnectionSlot()));
	} else {
		QLocalSocket socket;
		socket.connectToServer(m_unique);
		if (socket.state() == QLocalSocket::ConnectingState)
			socket.waitForConnected(200);
		if (socket.state() == QLocalSocket::ConnectedState) {
			socket.write(QCoreApplication::arguments().join(' ').toUtf8());
			socket.flush();
			socket.waitForBytesWritten(200);
		}
	}

	return is_server;
}


// Local server/shmem cleanup.
void qpwgraph_application::clearServer (void)
{
	if (m_server) {
		m_server->close();
		delete m_server;
		m_server = nullptr;
	}

	if (m_memory) {
		delete m_memory;
		m_memory = nullptr;
	}

	m_unique.clear();
}


// Local server connection slot.
void qpwgraph_application::newConnectionSlot (void)
{
	QLocalSocket *socket = m_server->nextPendingConnection();
	QObject::connect(socket,
		SIGNAL(readyRead()),
		SLOT(readyReadSlot()));
}


// Local server data-ready slot.
void qpwgraph_application::readyReadSlot (void)
{
	QLocalSocket *socket = qobject_cast<QLocalSocket *> (sender());
	if (socket) {
		const qint64 nread = socket->bytesAvailable();
		if (nread > 0) {
			const QByteArray data = socket->read(nread);
			// Parse and apply passed command-line arguments...
			qpwgraph_main *form = static_cast<qpwgraph_main *> (m_widget);
			if (form && parse_args(QString(data).split(' ')))
				form->apply_args(this);
			// Just make it always shows up fine...
			if (m_widget && !m_start_minimized) {
				m_widget->showNormal();
				m_widget->raise();
				m_widget->activateWindow();
			}
			// Reset server...
			setupServer();
		}
	}
}

#endif	// CONFIG_XUNIQUE


//----------------------------------------------------------------------------
// main.


int main ( int argc, char *argv[] )
{
	Q_INIT_RESOURCE(qpwgraph);

	qpwgraph_application app(argc, argv);

	if (!app.parse_args(app.arguments())) {
		app.quit();
		return 1;
	}

#ifdef CONFIG_XUNIQUE
	// Have another instance running?
	if (!app.setupServer()) {
		app.quit();
		return 2;
	}
#endif

	qpwgraph_main form;
	app.setMainWidget(&form);
	form.apply_args(&app);

	// Setup session manager shutdown (eg. logoff)...
	QObject::connect(
		&app, SIGNAL(commitDataRequest(QSessionManager&)),
		&form, SLOT(commitData(QSessionManager&)),
		Qt::DirectConnection);

	return app.exec();
}


// end of qpwgraph.cpp

